### ISSUE TITLE
Documents @ DataHub

### ISSUE THREAD

**Issue by julmot** [pimcore/data-hub#113]
After the Pimcore conference last week, I've heard that you need use cases, ideas and requirements for the [Documents integration](https://github.com/pimcore/data-hub/issues/49) as part of the DataHub. Therefore I'd like to share my two cents.

Our customers are currently relying on the Documents part of Pimcore for authoring and managing all of their non-dynamic contents – like e.g. landing pages, FAQ pages, contact pages, etc. They can currently create _multi language_ pages with around two dozen _bricks_. All of them can be managed using the _edit window_ (using a dedicated edit template). We have chosen this approach over classic WYSIWYG to avoid fields that wouldn't appear in the live website (like. e.g. a select field to choose the number of slides in a slideshow) and to avoid spending time in supporting Pimcore's HTML elements and classes that would manipulate our theme's HTML at least in edit mode (editables). Also, until DataHub is available in a stable version, we're rendering Twig templates directly from a frontend library, respectively including frontend templates with JSON data. In the context of a document, this means, that we're reading brick data (`.getData()`) and forwarding it to the frontend template itself. This approach is a temporary solution until we can stick to a classic decoupled approach using the GraphQl API – giving us the ability to use certain features of the decoupled approach in the meantime; the frontend has 100% control over the HTML and we don't have redundant HTML code (which would be the case when having a frontend library and separate backend integration layer). However, with this approach, we've experienced caching issues and would like to stick to GraphQl as soon as possible as we did this in the past with other systems.
Just mentioning this to let you know where not using classical WYSIWYG and this is not a use case for us.

From an end-user point of view we're building eCommerce systems where users can search, filter, browse, login, place and manage orders, manage profile information, etc. We're also having the use case of B2B resulting in a dedicated profile view for resellers.

We need the DataHub GraphQl API for our main application, so we won't have the "classic" approach with rendering templates backend-side beside of the decoupled/headless approach. Therefore, we need a solution that replaces the WYSIWYG approach.

What we would need when integrating documents as part of the GraphQl API are the following requirements:

- [ ] The editor wants to manage one or multiple bricks
- [ ] The editor wants to adjust the order in which bricks will appear
- [ ] The editor wants to edit bricks context-sensitive – at the place where it appears in the edit mode
- [ ] The editor wants to override certain information during the personalization process
- [ ] The editor wants to edit content for different languages
- [ ] The editor wants to preview a document with different viewports
- [ ] The user wants to edit documents using the existing documents tree that implicitly controls routing
- [ ] The end-user wants to be able to authenticate using the GraphQl endpoint
- [ ] A nice to have for the documents part of Pimcore anyway would be validation and constraints for specific fields

We've evaluated FieldCollections (thanks for @fashxp for this idea) as a workaround for the missing Documents integration. While this would allow editing one or multiple bricks, even with validation and multi-language support, we would probably miss personalization and routing features, as well as workflow options, multi-device previews and also content couldn't be authored context-sensitive anymore – only after the final save button is clicked and the preview tab has been opened.

In terms of a solution approach I personally see two options:
1. Headless: Implement it like in Drupal 8. You can author data and preview it afterward. Similar to FieldCollections but with built-in routing, multi-device previews, workflows and so on. This would fit all requirements except the one to edit content context-sensitive. It would mean to get rid of WYSIWYG at all – at least when the system is configured to do so. Since we've heard a lot about that multi-channel solutions will be the future during the conference, I think this would fit better to this philosophy because content can be managed regardless of the output channel
2. Decoupled: The application is integrated in some kind into Pimcore, allowing users to edit content context-sensitive, similar to WYSIWYG. It could be solved e.g. by embedding the decoupled application in the document ("Edit" and "Preview" tab could get merged). The decoupled frontend application could then [open an iframe in the edit window](https://github.com/pimcore/data-hub/issues/49#issuecomment-527786182) to author brick fields and call certain API endpoints to create/reposition/delete bricks (using e.g. a toolbar like it's implemented today). Since this would still allow Angular/React/Vue applications using GraphQl, it could be a good solution.

I'm happy to discuss this further and hear your thoughts.

**Comment by micru** [comment]
Hi, just wondering, are there any plans to implement the features as suggested by @julmot ? Our customers also rely very heavily on areabricks for documents and so do a lot of other Pimcore users i assume. For us the possibility to request data from areabricks via graphQL would be an absolute necessity, as we develop more and more single page JavaScript PWA

**Comment by julmot** [comment]
I've just re-checked the progress on this topic, seems like there's now basic support to [read documents](https://github.com/pimcore/data-hub/blob/master/doc/graphl/DocumentQueries.md) decoupled. But seems like support for bricks is missing and also the part of integrating decoupled apps into the WYSIWYG editor of Pimcore seems to be open.

**Comment by fashxp** [comment]
we have now quite generic integrations implemented for [document queries](https://pimcore.com/docs/data-hub/current/GraphQL/Query/Document_Queries.html) and [document mutations](https://pimcore.com/docs/data-hub/current/GraphQL/Mutation/Document_Mutations.html) that support all editable types of documents. 

so i'll close this one for now. input for further improvements is always welcome, please create new tickets based on the current implementation. 

the one thing we need to keep in mind in combination with graphql is though, that we need to be able to generate a very strict schema in advance - what turned out to be quite challenging in the document context ;-).

**Comment by julmot** [comment]
> we have now quite generic integrations implemented for document queries and document mutations that support all editable types of documents.
> so i'll close this one for now. 

I'm glad you made progress in this regard and you're supporting basic queries and mutations for documents!

However, as far as I can see you're missing the built-in integration of WYSIWYG editing. Have a look at the checkboxes I've posted last year. I don't think any of those is statisfied. Currently, the entire CMS area of Pimcore depends on Twig templates. Every brick has to come with a template to author and manage the content. But imagine when we don't have Twig templates anymore, as all our channels are decoupled using DataHub. Then, how do we authorize and manage content? We'd still have to ship Twig templates to let users edit their content with WYSIWYG, whereas we're consuming it without Twig. This doesn't make sense.

With the current API and with further work on it we'd probably be able to build a way to write content ourself using a mutation. Then we'd have both, a way to visualize the data using DataHub and manage the content with DataHub (mutation). But the entire WYSIWYG integration would be up to every project, which I feel is cumbersome and probably no one is interested in it, since it costs so much work. What we need is an integration of WYSIWYG with decoupled applications considering the checkboxes above.

**Comment by fashxp** [comment]
Do you have already an idea how to tackle these points?
I'm not sure, how this can be solved by datahub - as you challenge a couple of basic concepts of documents.

**Comment by julmot** [comment]
@fashxp Currently I see two options - I've mentioned them in my initial post here.

The second mentioned option would be a hybrid between WYSIWYG and completely form based authoring of content. Here, the decoupled application (or in other words every single project using DataHub) would only need to integrate an edit button and an overlay in the loook and feel of the application itself for each brick. In this overlay, we could simply embed an iframe that lists all the editables of this brick like it already does with an edit template currently. On save we could close and reload the brick of the page.

The application itself would be shown instead of the current preview tab in the CMS area and would detect the edit mode through a GET parameter in the URL and an existing admin user session.

What do you think?
